src/HOCON/HoconParser.cs

```diff
@@ -0,0 +1,1011 @@
+        {
+            var subField = sub.ParentField;
+
+            // first case, this substitution is a direct self-reference
+            if (sub.Path == subField.Path)
+            {
+                var parent = sub.Parent;
+                while (parent is HoconValue)
+                    parent = parent.Parent;
+
+                // Fail case
+                if (parent is HoconArray)
+                    throw new HoconException("Self-referencing substitution may not be declared within an array.");
+
+                // try to resolve substitution by looking backward in the field assignment stack
+                return subField.OlderValueThan(sub);
+            }
+
+            // second case, the substitution references a field child in the past
+            if (sub.Path.IsChildPathOf(subField.Path))
+            {
+                var olderValue = subField.OlderValueThan(sub);
+                if (olderValue.Type == HoconType.Object)
+                {
+                    var difLength = sub.Path.Count - subField.Path.Count;
+                    var deltaPath = sub.Path.SubPath(sub.Path.Count - difLength, difLength);
+
+                    var olderObject = olderValue.GetObject();
+                    if (olderObject.TryGetValue(deltaPath, out var innerValue))
+                    {
+                        return innerValue.Type == HoconType.Object ? innerValue : null;
+                    }
+                }
+            }
+
+            // Detect invalid parent-referencing substitution
+            if (subField.Path.IsChildPathOf(sub.Path))
+                throw new HoconException("Substitution may not reference one of its direct parents.");
+
+            // Detect invalid cyclic reference loop
+            if (IsValueCyclic(subField, sub))
+                throw new HoconException("A cyclic substitution loop is detected in the Hocon file.");
+
+            // third case, regular substitution
+            _root.GetObject().TryGetValue(sub.Path, out var field);
+            return field;
+        }
+
+        private bool IsValueCyclic(HoconField field, HoconSubstitution sub)
+        {
+            var subStack = new Stack<HoconSubstitution>();
+            subStack.Push(sub);
+            return IsValueCyclic(null, new List<HoconField> {field}, subStack);
+        }
+
+        private bool IsValueCyclic(HoconValue currentValue, List<HoconField> visitedFields, Stack<HoconSubstitution> pendingSubs)
+        {
+            while (pendingSubs.Count > 0)
+            {
+                var currentSub = pendingSubs.Pop();
+                if (!_root.GetObject().TryGetField(currentSub.Path, out var field))
+                    continue;
+                if (visitedFields.Contains(field))
+                {
+                    return true;
+                }
+                visitedFields.Add(field);
+                if (IsValueCyclic(field.Value, visitedFields, pendingSubs))
```

Comment:
Writing recursive functions in C# can be risky - they are not reduced. It's not a must, but maybe rewriting it into loop would be worth a while. Also keep the exit condition at the beginning.

Reply:
Done, unrolled the recursive method.